raft.go

```diff
@@ -692,8 +692,21 @@ func (r *Raft) leaderLoop() {
 				case err := <-doneCh:
 					if err != nil {
 						r.logger.Debug(err.Error())
+						future.respond(err)
+					} else {
+						// Wait for up to ElectionTimeout before flagging the
+						// leadership transfer as done and unblocking applies in
+						// the leaderLoop.
+						select {
+						case <-time.After(r.config().ElectionTimeout):
+							err := fmt.Errorf("leadership transfer timeout")
+							r.logger.Debug(err.Error())
+							future.respond(err)
+						case <-leftLeaderLoop:
+							r.logger.Debug("lost leadership during transfer (expected)")
```

Comment:
What do you think about just staying in the loop here instead of starting a _new_ `ElectionTimeout` ticker and having to duplicate the leftLeaderLoop code path too?

I was originally expecting that we'd not wait for _another_ `ElectionTimeout` but rather keep waiting for _up to_ the original one we started for the transfer to complete. I don't think it's terrible to do it this way but it strikes me that it's less code and _maybe_ simpler to reason able if the behaviour is just: block until leadership transfer works or one election timeout from handling the request...

The rationale for using ElectionTimeout here presumably was that, LeadershipTransfer is only an optimization to avoid just ungracefully stopping and having to wait for an election timeout... so letting leader transfer take longer than a whole election timeout is a little bit defeating the point and we should probably return fast and let the old leader just shut down if it wants to.

I think that rationale gets a bit more subtle in cases like Autopilot upgrade where we _rely_ on Leadership Transfer working before we move on to the next phase rather than just shutting down anyway.

tl;dr, I don't think the behaviour here is bad or wrong especially and it beats the bug without this wait. If it turns out easier to  follow to duplicate the code and make the behavior more explicit like you have here I'm OK with it. Just curious if you tried simply replacing this `else` branch with `continue` which I _think_ is also a correct fix?

Reply:
Yes, I tried that.  It got messy due to the other cases wanting to read from doneCh.  In the end I decided this was clearer.